Playlist block

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Adds a playlist and it's inner playlist track block.

Closes https://github.com/WordPress/gutenberg/issues/805

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The playlist block has been requested since 2017. Without the block, a playlist can be added by using the classic block, but that is overly complicated.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
For this block, I started with one media type, audio, rather than trying to build it for both audio and video because it would increase the complexity.

I pictured that the playlist should work similarly to the "classic" WordPress playlist.
Meaning:
- There would be only one audio element followed by an optional Tracklist with information like song title, artist, and track length.
- Clicking on a track in the Tracklist should play that track.
- When one track ends, the next track in the Tracklist should start.
- The block options should match the playlist options in the Media Library.

## Testing Instructions

Audio file for testing: https://wpthemetestdata.files.wordpress.com/2008/06/originaldixielandjazzbandwithalbernard-stlouisblues.mp3

Enable the experimental blocks option from the Gutenberg > Experiments plugin menu.
Upload a couple of audio files **through the Media Library.**
Edit the files in the Media Library and add title, artist and album to some files but not all.
Add a featured image to at least one file.

Insert the playlist block in the editor.

1. There should be a media placeholder from which you can open the media library and select audio files. Hold down Shift to select more than one file.
3. Confirm that all selected files are inserted in the block and that the first file is presented at the top of the block, above the tracklist.
4. Select the Edit button in the toolbar of the Playlist block . Upload or select a one or more files from the media library. Confirm that the files you choose **replaces** the current items in the block.
5. Please test the alignment options in the block toolbar.
6. Clicking on the play button of the audio element should be disabled.
7. Clicking on a playlist track should select the inner block and the block settings sidebar should show the options for the track, not the playlist itself. 
8. Confirm that you can select one or more playlist track blocks and move them up and down using the move down, move up, or drag options. 
9.  Confirm that there are no issues/errors when deleting or duplicating a playlist track block.
10. Select a  playlist track block and edit the title, album and artist by editing the text directly on the inner block in the canvas.
11. Select a  playlist track block and edit the title, album and artist by editing the text using the options in the sidebar.
12. Select a  playlist track block and delete all text contents. Confirm that placeholder texts are showing.
13. Select a  playlist track block and upload an album cover image.
14. Replace the album cover image.
15. Remove the album cover image.

16. Try the sidebar options: Select the parent playlist block.
- Show Tracklist -hides and shows the tracklist.
- Show artists in Tracklist -hides and shows artists in the tracklist if the artist info is available.
- Show number in Tracklist -hides and shows the number in the tracklist.
- Show images -hides and shows the image for the track.
- Order -change the order of the tracks.
- In the Styles tab, change colors, spacing and borders.

View the front of the website.
- Confirm that clicking the play button plays the track.
- When a track ends, the next track plays until you reach the last track.
- Clicking on a track in the tracklist switches the current track.
- The current track is marked in bold in the tracklist.
- Repeat this step with more than one playlist block placed on the same page.

### Screen Readers
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

#### Description of the block:
The block has three different sections.
At the top, there is an image, for example, an album cover, which is set to decorative with an empty alt attribute.
Next to the image is information about the current track: Title, album, and artist.
Below is a native audio HTML element with the standard controls for playing, pausing, volume, etc.
Below the audio element is the tracklist.
It is a list with buttons inside. Each list item and button represents a track and the button text is the song title, artist, and the length of the track.
The currently selected track: The one that is at the top of the block, has `aria-current=true.`
Activating a button selects that track.

#### Instructions
(I have only tested these steps with VoiceOver on mac)

Enable the experimental blocks option from the Gutenberg > Experiments plugin menu.
Upload a couple of audio files **through the Media Library.**
Edit the files in the Media Library and add title, artist and album to some but not all.
Add a featured image to at least oner file.

Add a playlist block in the block editor.
Focus is moved to an upload button in the block's placeholder.
Instructions about uploading an audio file or selecting one from the media library are announced.
Pressing the right arrow key or the tab key moves the focus from the Upload button to the Media Library button. Each button opens its respective file selection flow.
Confirm that it is possible to select multiple files.
When clicking on the button with the text "Select", the modal for the Media Library is closed.

After adding the audio files, the focus is moved away from the block to the editor canvas and one has to navigate to the block again to select it.  (I found this confusing, but learned that other media placeholders do the same)

Navigate to the block in the editor.
The block name is announced.
Pressing the down arrow key or the right arrow key moves focus to the audio element inside the playlist block.
The element type, the track title, the artist name, and the album name are announced if this information is available. Some tracks will only have a title. The same information that is announced is printed above the audio element, available to sighted users.

Pressing the spacebar or Enter key will play the track, and pressing the spacebar or Enter again will pause the track.
**Let me know if playing the track in the editor should be disabled.**

Pressing the down arrow key will move the focus to the button of the first track in the tracklist. The screen reader should announce the title, the band, the length of the track (if available), and the instructions "Select to play this track".
- Pressing the down arrow key now selects the richText element for editing the song title.
- Pressing the down arrow key now selects the second richText element, which is for editing the artist name.

Pressing the down arrow key again will select the next track in the tracklist, and so on, until you reach the last track.


##### Replacing or adding tracks 
Because the tracks are inner blocks, there will be two block toolbars where you can edit tracks.
From the block toolbar of the inner block, you can replace the individual, single track.
From the block toolbar of the parent block, the playlist, you can add multiple new tracks at once.

Press Shift + Tab to open the block toolbar.
Use the arrow keys to navigate to the button named "Edit"  or "Replace" respectively.
Activating the button opens a dropdown with two options: Open Media Library, and Upload.
Each option opens its respective file selection.
When the selection is complete, the screen reader announces that the media has been replaced.
(This is not accurate if you are adding a file, not replacing it, but I don't know if this can be solved?)
The focus is returned to the Edit button.

##### Block Sidebar
Both the playlist block and the inner track block has options in the block sidebar.

**The Playlist** has options both in the settings tab and styles tab.
Some of these options are visual, such as padding and margin and hiding the decorative image.
Optionally, test the first option in the settings tab, which is a toggle used to hide and show the tracklist:
With the block selected, press the tab key to move the focus to the sidebar.
Navigate to the block settings sidebar, past the Close button and the button that opens the Settings tab,
and the button that opens the Settings panel.
Find the checkbox option for "Show tracklist". Uncheck the checkbox.
Press the tab key repeatedly until you reach the "Skip to the selected block" button.
With the block selected, confirm that using the down arrow key does not move focus to the tracklist, since it has been turned off in the option.

**The track** has options in the settings tab.
With the block selected, press the tab key to move the focus to the sidebar.
Navigate to the block settings sidebar, past the Close button and the button that opens the Settings tab,
and the button that opens the Settings panel.
There will be three text input fields where you can update artist, album and title.
Below is an option for selecting an album cover image from the Media Library.

carolinan committed:
Basic playlist block

carolinan committed:
try to make some accessibility improvements

carolinan committed:
Park the PHP file until work can continue

carolinan committed:
Update index.php

carolinan committed:
Update index.php, try to solve merge conflicts

carolinan committed:
Remove the style variation

carolinan committed:
Remove the unused save, add escaping, update css

carolinan committed:
Update edit.js

carolinan committed:
Update edit.js

carolinan committed:
Update edit.js

carolinan committed:
fix CS issues

carolinan committed:
Fix two issues with the upload option.

carolinan committed:
Disable uploading and selecting multiple files in the replacement flow.

carolinan committed:
Update the placeholder instructions

carolinan commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,342 @@
+									src={
+										ids[ trackListIndex ]?.image.src
+											? ids[ trackListIndex ]?.image.src
+											: '/wp-includes/images/media/audio.png'
```
I need some help with coming up with a better way to add the placeholder image.
This image is missing when the file is uploaded and not selected from the Media Library.

carolinan commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,342 @@
+												ids[ trackListIndex ]?.title
+											),
+										} }
+									/>
```
`dangerouslySetInnerHTML`, `safeHTML` are needed because the Media Library and the "classic" playlist allows using some HTML here, like making the text bold.

carolinan committed:
Correct a translators comment.

carolinan committed:
Hide tracklist options if the tracklist itself is toggled off

carolinan commented:
I am keeping this is a draft because I am seeing a problem where when you add the block and select more than one file from the media library, the track information with the artist, album and length is only added for one track, not all.
While if I use the media replacement flow (edit), the track information for all tracks is updated.

jasmussen commented:
Thanks for the ping. Nice! Works well. Here's a GIF showing the placeholder, the inspector options, selecting items:

![playlist](https://github.com/WordPress/gutenberg/assets/1204802/f31ce684-ed9b-4e1e-8252-45e214c5ed96)

There are a couple of small things:

* Can probably add the border control to the inspector, rather than outputting a CSS file just for this block.
* In fact ideally we can remove as much of this CSS as possible, and rely on core block editor features. Both padding, block spacing, margin, border, radius, background, foreground colors come to mind, that would afford a great deal of pattern extensibility.
* Not sure if it's a box-sizing thing, but there's a curious behavior when applying padding to this one, moving it outwards instead of inwards, which is unlike how the Audio block works.

Quick instinct, this is not far from being able to work well. What do you think about next steps? Should I reopen the issue I just closed? Should we open a new one? Do you think it fits better as a canonical block? It may be a perfect candidate for a canonical block based on my gut instinct that fewer people in these days upload audio files and add them to playlists.

carolinan commented:
It does not need to be prioritised for this release even, I wanted to show that it was not completely abandoned :)

jasmussen commented:
In any case, I'm subscribed here, happy to help!

carolinan commented:
Is there room to pick this up for 6.7 or 6.8? I saw that there were discussions around styling the audio block.

carolinan committed:
Try to fix CS issues, update to apiVersion 3

carolinan committed:
Add since

carolinan committed:
Add color block supports

jasmussen commented:
I think the next step might be to ping @WordPress/gutenberg-core, perhaps.

creativecoder commented:
I've added [a new flag on the experiments page](https://github.com/WordPress/gutenberg/pull/64121) for Experimental blocks, because I don't think we need a separate experiment for each block (though I left the form blocks experiment in place, for now). Would you like to use it for the Playlist block?

I'm going to be AFK for a few weeks, but will plan to take a closer look at this PR when I get back later in August!

carolinan committed:
Add __next40pxDefaultSize and __nextHasNoMarginBottom to various controls .

carolinan committed:
Use the option for the experimental blocks.

carolinan committed:
WIP: Use the interactivity API.

carolinan commented:
I have started replacing the old view.js file using the interactivity API, but I am still learning.
I have not started making updates to the design to match the proposal in the issue.

carolinan committed:
Build the core-blocks.md doc to update the block supports

gziolo commented on packages/block-library/src/playlist/block.json:
```diff
@@ -0,0 +1,70 @@
+{
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 3,
```
The following change is necessary so this block doesn't get synced to WordPress core as long as it is experimental:
```suggestion
	"apiVersion": 3,
	"__experimental": true,
```

It's used with other blocks like the Form block:

https://github.com/WordPress/gutenberg/blob/67fca2ebcc9f0caafccbcaef15021d23cb0bf3dd/packages/block-library/src/form/block.json#L4

Related docs:

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/README.md#adding-new-blocks

gziolo commented on packages/block-library/src/playlist/index.php:
```diff
@@ -0,0 +1,146 @@
+	$current_album      = isset( $attributes['ids'][0]['album'] ) ? $attributes['ids'][0]['album'] : '';
+	$current_artist     = isset( $attributes['ids'][0]['artist'] ) ? $attributes['ids'][0]['artist'] : '';
+	$wrapper_attributes = get_block_wrapper_attributes();
+	$placeholder_image  = '/wp-includes/images/media/audio.png';
```
Will this URL work correctly for sites that are served from a subfolder of the domain? Example: https://my-site.com/subdfolder/.

carolinan replied:
Probably not.
The new design does not show a placeholder image, so this should probably be removed, or replaced with an SVG.

carolinan replied:
I have removed this iamge.

carolinan committed:
Index.php: Re-add the conditions for the block options.

carolinan committed:
Index.php: update spacing

carolinan committed:
Playlist: Add caption

carolinan committed:
Re-build docs

carolinan committed:
Re-add the autoplay of the next track when one track ends.

carolinan committed:
Remove the track title from the current track

carolinan committed:
WIP: Update styles

carolinan committed:
Update style.scss

carolinan committed:
Add placeholder artist and album, remove placeholder image

carolinan committed:
Update colors to work better when the user selects text and background color

ramonjd commented:
Thank you for working on this @carolinan 

I think it's a great idea and from what I've tested, the implementation is working well.

Just a couple of thoughts, feel free to ignore them.

What do you think about an "add track" button in the editor?

<img width="300" alt="Screenshot 2024-11-06 at 10 54 12 am" src="https://github.com/user-attachments/assets/6545e66a-fe78-40fd-b8a4-fe1a4f549051">

Another nice iteration would also be to be able to move tracks around in the list and delete them in place.

Also, I think it would be clearer to have the Track name as the main title, with the artist and album as the sub titles.

| Before | After |
|--------|--------|
| <img width="300" alt="Screenshot 2024-11-06 at 10 45 53 am" src="https://github.com/user-attachments/assets/ec562180-8d43-4a67-9495-1bbbc824cf80"> | <img width="300" alt="Screenshot 2024-11-06 at 10 54 07 am" src="https://github.com/user-attachments/assets/fdf18035-da8c-4f42-b613-2b683cbe0cfe"> |

carolinan commented:
To manage the tracks individually, they would need to be inner blocks. Either the audio block with several modifications, or a new block. This is mentioned in the PR description.

I did not include this because the repeated feedback I get on my blocks is that they are not simple enough (like the details block).

talldan commented:
> I did not include this because the repeated feedback I get on my blocks is that they are not simple enough (like the details block).

I think that Details block feedback was partly from me, but it was only relevant to the Details block, which had more than one level of inner blocks in that iteration and was tricky to edit because of it.

This seems like a better use case, as a playlist is a close equivalent to a gallery (also uses inner blocks), and it's a single level of inner blocks being discussed. I'd be curious for the thoughts of others too.

carolinan committed:
WIP: Use inner blocks for the tracks

Not ready for testing.

carolinan committed:
Track: Use "Replace" instead of "Edit" in the toolbar

carolinan committed:
Track: Set the text align on the button to left.

carolinan committed:
Track: Update block supports

carolinan committed:
Playlist: Try showing the inserter at the bottom of the block.

ramonjd commented:
Just dropping by to test the progress, as I'm excited about this block!

The innerblocks are feeling really natural to me. I love that I can reorder the tracks too. ❤️ 

https://github.com/user-attachments/assets/1fd67180-372f-4ae9-a0fb-2430b10282b8

carolinan committed:
Adjust the track button CSS class and move the appender

carolinan committed:
Fix the invalid HTML of the list on the front

carolinan committed:
Track: update block supports

carolinan committed:
WIP: Prevent duplicate tracks

This is only a partial solution. Users can still add duplicate tracks by replacing existing inner blocks.

There is some repetition of code that needs to be solved.

carolinan committed:
Move the tracklist inside `<Disabled>`

carolinan committed:
Playlist: Rename the images attribute to showImages

carolinan committed:
Revert placing the tracklist within Disabled since it prevented adding new tracks.

carolinan committed:
Revert the prevention of duplicate tracks

Based on feedback, there may be scenarios where users want to use duplicate tracks. For example to play commercial jingles between songs / episodes etc.

carolinan committed:
Playlist: rename the tracklist block attribute to showTracklist

carolinan committed:
Playlist: Update the tracks block attribute when the user changes position of the tracks.

carolinan committed:
Playlist: Fix the display of the album image

carolinan committed:
Re-add the track title to the current track.

carolinan committed:
Try to fix a JS error when a track is re-positioned.

carolinan committed:
Playlist: Add the current title as a data-wp-text

carolinan committed:
Playlist: Correct the position of the artist and album for the current track.

carolinan committed:
Track: Try to improve responsiveness

carolinan committed:
First iteration of adding the album cover image to the track

This code is largely copied from the Video block Poster image feature.

carolinan committed:
Adjust the CSS

carolinan committed:
Update the monitoring of changes to inner blocks

carolinan committed:
Update how the current track is managed

Add a new block attribute called "currentTrack", and pass it as a context from the parent to the inner block.

Set the id (the post id of the audio media file) as the value of currentTrack:
- On file selection
- On removal of the inner block
- On re-ordering of the inner blocks

To do: It would be good to reduce the code duplication.

carolinan committed:
Update the appender, prevent grouping inner blocks

carolinan committed:
Prevent PHP notices when the current track is the media placeholder

This solution is not ideal, nor complete.
The placeholder where you select new tracks should not be able to be the "current track".

carolinan committed:
Move the current track to a new constant

carolinan committed:
Adjust the styling of the placeholder, removing the border

carolinan committed:
Display a placeholder for the current track

If the current track is the media placeholder (from the block inserter), then display a placeholder title, artist and album, and remove the source from the audio element.

carolinan committed:
Update how the current track is handled

Try to safe guard from invalid current track id's.

carolinan committed:
Remove a constant that was only used once

carolinan committed:
Add a comment about sanitizing HTML in data from the media library

carolinan committed:
Add checks for the media_details attachment data

Add checks for the media_details fields for artist, album and length, to populate the track data when the upload feature is used.

carolinan committed:
Playlist track: Update the MediaReplaceFlow

In the MediaReplaceFlow:
- Remove unused "addToPlaylist" parameter.
- Replace mediaIds with mediaId since the track works with singular files.
This ensures that when the user selects "Replace" in the block toolbar and opens the MediaLibrary, the current track is pre-selected.

carolinan committed:
Playlist: Update MediaReplaceFlow

Re-add the mediaIds to the MediaReplaceFlow, so that when the user clicks on "Edit" in the toolbar and opens the Media Library, the current tracks are pre-selected.

carolinan commented:
Im having problems solving the large amount of undo steps, I am reading up on past issues related to this.

Two changes I made to day also resulted in tracks being duplicated during the replace flow. This is known and a work in progress.

carolinan committed:
Playlist: Try to reduce the number of undo steps

carolinan committed:
Playlist MediaReplaceFlow: remove "value"

I don't recall what I was trying to do here, so I am removing the custom prop.

carolinan committed:
Track: introduce blob management and clean up the MediaReplaceFlow

The purpose of this change is to improve the user experience when uploading new files and when undoing the insertion of new tracks.
This is copied from recent changes made to the audio and video blocks.

carolinan committed:
Playlist: Try to solve the duplcation of existing tracks on file upload in the media replacement flow

carolinan committed:
WIP: Make the block actually dynamic

Reduce the stored block attributes and use the data from the database.
+ Code reduction

This is a WIP and not fully functional.

carolinan committed:
WIP: Add the markup for the current track at the top of the playlist

Create the HTML markup, including the interactivity API state, and insert it inside the block's figure element.

carolinan commented:
Any way, just noting for anyone dropping by that this is not testable at the moment.

draganescu commented:
Hi @carolinan there is so much to do 😂 

>update the Rest API to allow users (with the correct capability) to edit the attachment data.

I am very interested in landing a version of this block! I'll try to help as much as i can. Ideally we should ship the least possible version and then iterate. I'll do a review here ASAP 🚀

ramonjd commented:
Great work so far. Thanks for pursuing this block. I love it.

> I am not sure how to move forward with this because I don't have the knowledge to update the Rest API to allow users (with the correct capability) to edit the attachment data. That is the (post) title, featured image (poster / album cover art) and the media details: artist and album.

I'm just trying to gather ideas of what makes sense in terms of permissions, and dump some ideas.

What do you think are the requirements? 

- Should users that can edit the post containing a play list block be able to edit the attachment's post data, regardless of whether they uploaded the attachment?
- I'm assuming there should be a limit on the fields post editors can update, as you mention, title, details etc. Are there any new fields required?
- Does it make sense to limit these permissions to attachments associated with the [current post only](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php#L1026)? Or go even further and (somehow) limit it again to the playlist block?

From a quick check, the attachments post type has [upload files capabilities](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/post.php#L89), so possibly it'd involve tweaking that via [register_post_type_args](https://developer.wordpress.org/reference/hooks/register_post_type_args/).

Since I expect more fine-grained control is required, it would involve extending `WP_REST_Attachments_Controller` methods to handle the permission changes. 

The attachments controller inherits its parent post controller's `update_item_permissions_check`. That method [tests edit permissions](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L859).

Sorry for all the questions and wall of text! I'm thinking it could be a separate PR. I'm also happy to help out when I can scrape the time together.

Keen to hear what @draganescu thinks too.

carolinan commented:
As far as I know, the "post parent" of an attachment can only be one post, and it is the post that was being edited when the audio file was uploaded. So I don't think that connection can be used.

carolinan commented:
- Should users that can edit the post containing a play list block be able to edit the attachment's post data, regardless of whether they uploaded the attachment?

Yes. The problem is that when you upload an audio file, sometimes there is no meta data. Things like the album name might not exist. The title may be the file name, as one word. Not being able to edit those detail from the block would be a subpar experience.
The worst step is trying to add an album cover, since you can't do that unless you **exit the post you are editing,** open the Media Library from the WordPress admin, locate the item, select "Edit more details" and set a featured image to the attachment post.

- I'm assuming there should be a limit on the fields post editors can update, as you mention, title, details etc. Are there any new fields required?

I don't think there are any new fields required. The options should be limited to the visible text content. For example, it would not make sense to have an option to update the bit rate or file length. 
- Title (this is awkward, because the attachments have more than one! the post_title and the media details title)
- Artist
- Album
- Genre? Genre exist but it is not used in the block design.

And these are just if the block is limited to being a playlist for audio, not video.

talldan commented:
I guess it makes most sense to stick pretty closely to how the media library already works.

It might be worth pinging some media maintainers for advice on which fields to update.

I had a quick look, and it looks like something like this works very basically for editing the title of the media:
```jsx
const [ title, setTitle ] = useEntityProp( 'root', 'media', 'title', id );

// ...

<TextControl
	__nextHasNoMarginBottom
	__next40pxDefaultSize
	label={ __( 'Title' ) }
	value={ title }
	onChange={ ( titleValue ) => {
		setTitle( titleValue );
	} }
/>
```

But I'm not aware of all the permutations, as you say there's also the `media_details` property which contains another title. When I edit in the media library, only the post title seems to change. Editing the artist works differently in that it changes the `media_details.artist`. Again though, I'm not sure if this is consistent for all audio media.

`useEntityProp` doesn't work for nested properties (e.g. `media_details.artist`), but I expect `useEntityProp( 'root', 'media', 'media_details', id )` would work, though would have to be careful not to delete any existing properties when updating (they'll need to be merged into the object I expect).

To check for capabilities, `select( coreStore ).canUser( 'update', { kind: 'root', name: 'media', id } )` should work. Blocks like Post Title have example code.

carolinan commented:
Thank you, I was trying to update the attachment post type, I will try root media shortly.

carolinan commented:
About the titles, I would have liked to use the media details title, but in the Media Library it is the post title field that is editable, so
I think it would be too confusing: If you went into the Media Library and updated the title field, the block would not reflect the change.

talldan commented:
I think the `WP_REST_Attachments_Controller` handles all the `/wp/v2/media` routes (which is what the `media` entity in gutenberg calls through to). At least there's no `WP_REST_Media_Controller`, and in the docs there's no `attachment` endpoint listed.

Unfortunately, looking at the REST API documentation it seems as though`media_details` is read only, so it might not be possible to update it from the block:
https://developer.wordpress.org/rest-api/reference/media/

I'm not sure if there's a reason for not allowing updates via the REST API, or if it's unimplemented. 🤔

carolinan commented:
It would be good to know the reason why they are read only before continuing.

Also, can we update the message that is shown when saving?
When I change the title and save, it says Media "Untitled" which can be confusing.
Perhaps this should not be optional? Is there a way to disable the checkbox?
![The save panel in the post editor, showing that there is a change to Media, but the label of the checkbox says "Untitled"](https://github.com/user-attachments/assets/cd66548e-1451-43c7-be24-549295ba061a)

carolinan committed:
Playlist track: Save changes to the title of the track using useEntityProp.

draganescu commented:
> using the attributes was that it is too much data,

@carolinan I can see why the trimmed down markup is less prone to breaking, but I would argue that it's not a concern strong engough to guide a whole lot of architecting _around_ it. 

> too much data [...] increases the risk of the block breaking by user error.

Is this about manual editing of block markup? Because if it is, that should not be something we can even prevent.

carolinan commented:
Ok I will try to rephrase.

The main reason why I made it dynamic is because it is my opinion that if I update an attachment it should have the same content everywhere. I did not add that to the other comment because it had already been brought up.
The amount of data and the code reduction is just one more point.
And because the block already uses a PHP file to enqueue the view file and add the data-wp attributes and state , it is well suited for being dynamic.

luisherranz commented on packages/block-library/src/playlist/index.php:
```diff
@@ -0,0 +1,127 @@
+
+	// Add the markup for the current track
+	$html = '<div><div class="wp-block-playlist__current-item">';
+	if ( $show_images && $current_image ) {
```
We always need to add the IMG element; otherwise, it will only be added if the first track contains an image. But if the first track doesn't contain an image and the following tracks contain images, those images won't be visible.

carolinan replied:
Displaying an empty image tag is not a good practice.

luisherranz replied:
It is hidden when there is no image.

`data-wp-bind--hidden="!state.currentTrack.image"`

luisherranz commented on packages/block-library/src/playlist/index.php:
```diff
@@ -0,0 +1,127 @@
+		$html .=
+		'<img
+			class="wp-block-playlist__item-image"
+			src={ ' . esc_url( $current_image ) . '}
```
The [server directive processing](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/core-concepts/server-side-rendering/) of the Interactivity API takes care of adding the attributes to the HTML. There's no need to add them manually.

```suggestion
```

(This applies to other places in the code, not just here)

luisherranz commented on packages/block-library/src/playlist/index.php:
```diff
@@ -0,0 +1,127 @@
+		'<img
+			class="wp-block-playlist__item-image"
+			src={ ' . esc_url( $current_image ) . '}
+			alt=""
```
We should add an `alt` attribute, shouldn't we?

carolinan replied:
I consider it decorative. It would be a very generic alt attribute otherwise, which would be the same for every image, that would not bring much value, such as "Track poster image" or similar.

luisherranz commented on packages/block-library/src/playlist/index.php:
```diff
@@ -0,0 +1,127 @@
+		data-wp-watch="callbacks.init"
+		"/></audio></div>';
+
+	$figure = null;
+	preg_match( '/<figure[^>]*>/', $content, $figure );
+	if ( ! empty( $figure[0] ) ) {
+		$content = preg_replace( '/(<figure[^>]*>)/', '$1' . $html, $content, 1 );
+	}
```
Why do you need to add `figure` conditionally? In what cases can it exist?

carolinan replied:
I am not trying to add a figure.
I am finding the figure to insert the HTML for the current track inside the figure.

There are two reasons:
- Because I was in the process of making the block dynamic, the save.js did not have the same access to the information, and did not include the markup for the current track, so I had to add it to the PHP file.

- Even before this change, when I was using the HTML API to add the wp-data attributes, the HTML API would add the attributes to the wrong HTML element. The title ended up as the album etc.
I could probably have solved this with some more time, but because of the mentioned change, I did not need to continue working on it at that point.

luisherranz reviewed (commented):
I've only inspected the frontend code, including the generated PHP and the `view.js` file. Great job 🙂

Using the HTML API to read the iAPI context of the tracks is very clever; we hadn't thought of it until now! Maybe we'll use that pattern in the Gallery block as well.

A few comments:

- I've noticed that you've used some imperative logic to control the frontend and in general, it's better to do everything in a [declarative manner](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/core-concepts/the-reactive-and-declarative-mindset/) as it leads to less bugs.
- It seems there's a bug that always selects the first song even if you click on the others. I'm not sure how I ended up in that unstable state, nor how to fix it.

https://github.com/user-attachments/assets/0440faf2-269f-4b6b-a31e-3f94438523f1

- To fix this bug and make things more declarative, it would be good to generate an array containing unique IDs for each track on the server, which can be used to know which audio is playing, which one is next, etc. 
- It also seems there’s a bug when selecting an image in the editor. I can't select an image and have it appear on the frontend.

https://github.com/user-attachments/assets/c5d9c1c5-ef4f-4c48-ab64-96acd6e00b98

- I've also noticed that the `aria-label` doesn't change when the Track changes. I understand that this is a bug, right?

I've made a small refactoring to try to fix these issues and implement the populated server array.

luisherranz committed:
Remove imperative code and add server array

luisherranz commented on packages/block-library/src/playlist-track/index.php:
```diff
@@ -0,0 +1,104 @@
+	$media_id           = $attributes['id'];
+	$attachment_meta    = wp_get_attachment_metadata( $media_id );
+	$wrapper_attributes = get_block_wrapper_attributes();
+	$url                = wp_get_attachment_url( $media_id );
```
I switched to `wp_get_attachment_url` because `$attachment_meta['url']` was giving me problems, but feel free to revert.

luisherranz commented on packages/block-library/src/playlist-track/index.php:
```diff
@@ -0,0 +1,104 @@
+	$aria_label         = $title;
+
+	if ( $title && $artist && $album ) {
+		$aria_label = sprintf(
```
I've moved the Aria Label to the tracks to include it in the state and be able to change it dynamically.

luisherranz commented on packages/block-library/src/playlist-track/index.php:
```diff
@@ -0,0 +1,104 @@
+	wp_interactivity_state(
+		'core/playlist',
+		array(
+			'tracks' => array(
```
Now there's an array in the state that includes all the information about each track. This makes it easier to access it from anywhere.

carolinan replied:
It would be good to have an example in the documentation.

luisherranz replied:
Yes, we plan to talk about this type of patterns in a new section of [Core Concepts](https://github.com/WordPress/gutenberg/issues/62921).  Right now, it’s described as Data Flow Patterns, but we’ll try to find a better name.

luisherranz committed:
Add missing derived state in the server

luisherranz commented on packages/block-library/src/playlist/index.php:
```diff
@@ -0,0 +1,137 @@
+		)
+	);
+
+	// Finds the unique id of the current track and populates the playlist array.
+	$p               = new WP_HTML_Tag_Processor( $content );
+	$playlist_tracks = array();
+	while ( $p->next_tag( 'button' ) ) {
+		$track_context     = $p->get_attribute( 'data-wp-context' );
+		$track_unique_id   = json_decode( $track_context, true )['id'];
+		$state             = wp_interactivity_state( 'core/playlist' );
+		$playlist_tracks[] = $track_unique_id;
+		if (
+			isset( $state['tracks'][ $track_unique_id ]['media_id'] ) &&
+			$state['tracks'][ $track_unique_id ]['media_id'] === $current_media_id
+		) {
+			$current_unique_id = $track_unique_id;
+		}
+	}
```
I'm using your pattern to populate the array with all the tracks and the current track.

I haven't put any guard against undefined/null values, so it would be great if you could take a look. Thanks!

luisherranz commented on packages/block-library/src/playlist/view.js:
```diff
@@ -0,0 +1,60 @@
+			},
+		},
+		callbacks: {
+			autoPlay() {
+				const context = getContext();
+				const { ref } = getElement();
+				if ( context.currentId && context.isPlaying ) {
+					ref.play();
+				}
+			},
```
This callback makes it so that if there's a song playing and another one is clicked, it will also start playing automatically. I’m not sure if it’s something you had or not. I think you did.

carolinan replied:
The only thing is that there used to be a pause when the track was switched, because the experience of immediately starting to play the next song was a little unpleasant.

luisherranz reviewed (commented):
@carolinan It would be great if you could take a look at this refactoring, as I'm not sure if I've modified the behavior in any way.

Again, great work, this is a pretty cool block 🙂

PS: I haven't checked the communication between different playlist blocks. For example, if you play one, it should pause another. However, it's something that wouldn't be difficult to add.

carolinan commented:
Thank you for the review  @luisherranz. The image is not implemented, hence
> Any way, just noting for anyone dropping by that this is not testable at the moment.

As part of making the block dynamic I wanted to save the image to the attachment, and I had not started.

carolinan commented:
It becomes confusing for me when I have a few people asking me to revert parts that make the block dynamic and different people making changes to the code that well, makes it _more_ dynamic.

_With "dynamic" I am mean the part where the block is fetching the data from the attachment at all times, instead of using block attributes.

draganescu commented:
> It becomes confusing for me when I have a few people asking me to revert parts that make the block dynamic and different people making changes to the code that well, makes it more dynamic.

@carolinan it is confusing! We're all [sharing arguments](https://github.com/WordPress/gutenberg/pull/50664#issuecomment-2504455942) I hope - the point is not _to merge_ is to solve the problem in the best way we can today.

@luisherranz I am arguing that we make the block not ask for attachment data on the front end but store track data as local attributes. My argument is [here](https://github.com/WordPress/gutenberg/pull/50664#issuecomment-2500220564) and [here](https://github.com/WordPress/gutenberg/pull/50664#issuecomment-2504455942), basically:

- we dont have a clear way to expose that media library edits affect data in pages created with blocks
- we don't have a clear way to expose that data edited in block UI affects attachment entities in media library

Do you think that if the playlist block's front end behaviour is only related to its UI but the data comes from block attributes it is wrong?

carolinan commented:
> * To fix this bug and make things more declarative, it would be good to generate an array containing unique IDs for each track on the server, which can be used to know which audio is playing, which one is next, etc.

One interesting part of this that, for a time during development, every track needed to be unique: you couldn't add it more than once, and the post id for the attachment could be used for this.
But during a peer review this was an undesired behavior. The example that was brought up was that, if you are adding a list of podcast episodes, you may want to repeat a track that has advertising, between the episodes.

luisherranz commented:
> Do you think that if the playlist block's front end behaviour is only related to its UI but the data comes from block attributes it is wrong?

I don't have a strong opinion about that (you'll probably have better luck asking other people), but I wonder: what does saving audio metadata as attributes have to do with whether the block is static or dynamic?

draganescu commented:
> what does saving audio metadata as attributes have to do with whether the block is static or dynamic?

It not a direct result, but if the block is dynamic _because_ we need to save and get data from attachment entities on render, then having the data in block markup makes the need to be dynamic not a need anymore.

luisherranz commented:
Ok, thanks for the clarification.

If you ever end up making this a static block, please do not serialize the directives of the Interactivity API. Inject them dynamically using the HTML API.

carolinan commented:
> Ok, thanks for the clarification.
> 
> If you ever end up making this a static block, please do not serialize the directives of the Interactivity API. Inject them dynamically using the HTML API.

Can you re-write that in plain English?

carolinan commented:
Since there has not been any other input on the mechanics of the block for the past 5 days, I will continue to re-add the attributes.
- The Media Library will only provide initial data on file selection.

luisherranz commented:
> Can you re-write that in plain English?

What I mean is that if you finally convert this block into a static block, please don't add the directives to the content stored in the database (this is commonly known as "serializing"). Instead of doing that, use a `render_block` filter to add them dynamically using the HTML Tag Processor.

carolinan committed:
WIP: re-add the block attributes

carolinan commented:
I am still not sure I understand.

---
I have pushed an update but I will not be able to continue until after the new year.
There are several known issues that are not documented and the PR description and video is not up to date.

luisherranz commented:
> I am still not sure I understand.

Imagine that you have this dynamic block (simplified).

```html
<div data-wp-interactive="myPlugin" data-wp-init="actions.someInitialization">
  <button data-wp-on--click="actions.someHandler" data-wp-text="state.buttonText"></button>
</div>
```

Theoretically, if you want to move it to a static block, you could almost copy and paste the markup in the `save` function:

```js
const save = () => (
  <div data-wp-interactive="myPlugin" data-wp-init="actions.someInitialization">
    <button data-wp-on--click="actions.someHandler" data-wp-text="state.buttonText"></button>
  </div>
);
```

But if you do that, there will have to be deprecations every time a change o refactoring is made to the frontend implementation.

For that reason, instead of that, I recommend not storing the directives in the database, meaning they shouldn't exist in the `save` function.

carolinan commented:
Thank you.
Well is it correct now?
Why does it have to be a filter and not the callback in register_block_type_from_metadata?

luisherranz commented:
Oh, sure. You can use the render callback as well 🙂

carolinan committed:
Play the track that is being switched to.

carolinan commented:
A list of some known bugs:

- It is not possible to move the tracks with the toolbar controls when more than one track is selected.
- When two instances of the same tracks are placed in the playlist, and one is the current track, then the editor and front shows different tracks as the current track. In the editor, it is the first of the two tracks. On the front is is the last.
- When two instances of the same tracks are placed in the playlist, the editor shows both items with a bold text style, which is a style used to indicate which track is the current track. On the front, only the actual current track uses bold.

carolinan committed:
Add a placeholder for missing track titles on the front of the site

carolinan committed:
Try to implement the unique id

carolinan committed:
Update index.php

carolinan committed:
Try removing the tracks block attribute and only use the inner blocks

carolinan committed:
Let long titles wrap and not overflow the container

carolinan committed:
remove test code

carolinan committed:
Ensure that each inner block has a unique ID even when duplicated

carolinan committed:
Don't include the media placeholder in the tracklist

carolinan committed:
Fix the options to hide the tracklist and artist name

carolinan committed:
Tracklist: replace innerBlocksProps with innerBlocksProps.children

Use children instead, as the props on the tag itself broke the border styles.

carolinan committed:
Re-add the early return.

carolinan committed:
Fix the CSS for hidden tracklist

carolinan committed:
Update style.scss

carolinan committed:
Update since to 6..9.0

carolinan committed:
Playlist track: correct the data-wp-context in the editor

carolinan committed:
Try to isolate each playlist block in the global state by using a unique id as a prefix

carolinan commented:
I had to change how the block uses the interactivity API `global state`. 
When a playlist block was duplicated, each track's  unique id was no longer truly unique, and this caused a problem with the album cover image.
The image that was used on the playlist block that was placed last in the content, was also used in all other playlist blocks
(as long as the "show image" option was enabled).

It still uses the state, but, how do I say this in plain English,.. each playlist is isolated because the track data is nested one level deeper under a playlist id.

MaggieCabrera reviewed (approved):
This is testing really nicely for me, since it's under an experiment I think it's fair to bring it in and we can iterate if we feel like we need to

scruffian commented on packages/block-library/src/playlist-track/edit.js:
```diff
@@ -0,0 +1,295 @@
+				<MediaReplaceFlow
+					name={ __( 'Replace' ) }
+					onSelect={ onSelectTrack }
+					accept="audio/*"
```
What's the difference between `accept` and `allowedTypes`?

scruffian replied:
`accept` is the native browser file picker filter                                                                                                                                                      `allowedTypes` is the WordPress Media Library filter

MaggieCabrera commented:
I might have jumped the gun :D I get an error when trying to change the sort order and it doesn't work: `Uncaught TypeError: can't access property "localeCompare", b.attributes.uniqueId is undefined`

mikachan commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,469 @@
+
+	// Ensure that each inner block has a unique ID,
+	// even if a track is duplicated.
+	useEffect( () => {
```
I think this could be fixed in a follow-up, but it looks like this useEffect is running on every render. We could try memoising the unique IDs to avoid unnecessary re-renders.

Something like:

```
const uniqueIds = useMemo( () => {
	return innerBlockTracks.map( ( block ) => block.attributes.uniqueId );
}, [ innerBlockTracks ] );
```

And then add `uniqueIds` as a dependency to the useEffect.

carolinan replied:
I have tried many different alternatives but was not able to get it working :(

mikachan replied:
I think it's OK to come back to this later, don't let it block this PR!

scruffian commented:
When the block saves it outputs an `<ol>` in the markup - I wonder if it would be better to omit the `<ol>`, as it doesn't add much useful semantic information, and it will make it more complex to update the markup in future.

scruffian reviewed (approved):
Thanks for working on this, I love it ❤️ 🎧 . I think we should bring this in and keep iterating.

getdave commented on packages/block-library/src/playlist-track/index.php:
```diff
@@ -0,0 +1,74 @@
+		sprintf(
+			/* translators: %s: track length in minutes:seconds */
+			'<span class="screen-reader-text">' . esc_html__( 'Length:' ) . ' </span>%s',
+			$length
```
```suggestion
			esc_html( $length )
```

scruffian replied:
Not relevant anymore since the whole translation is escaped.

getdave commented on packages/block-library/src/playlist/index.php:
```diff
@@ -0,0 +1,173 @@
+					);
+				}
+
+				$tracks_data[ $unique_id ] = array(
+					'url'       => $url,
+					'title'     => $title,
+					'artist'    => $artist,
+					'album'     => $album,
+					'image'     => $image,
+					'ariaLabel' => $aria_label,
+				);
```
These values may need escaping. 

For example `ariaLabel` is used in `data-wp-bind--aria-label` so will need `esc_attr`.

scruffian replied:
Done in 9e32875284b

getdave reviewed (commented):
Couple of things to consider prior to merge.

t-hamano commented on lib/blocks.php:
```diff
@@ -88,6 +90,8 @@ function gutenberg_reregister_core_block_types() {
 				'post-author-name.php'             => 'core/post-author-name',
 				'post-author-biography.php'        => 'core/post-author-biography',
+				'playlist.php'                     => 'core/playlist',
+				'playlist-track.php'               => 'core/playlist-track',
```
Let's put them in strict alphabetical order (directly under patterns.php).

t-hamano commented on packages/block-library/src/playlist/save.js:
```diff
@@ -0,0 +1,46 @@
+					( ! showArtists
+						? ' wp-block-playlist__tracklist-artist-is-hidden'
+						: '' )
+				}
```
Let's avoid concatenating class names as strings, as this can introduce unintended spaces.
```suggestion
				className={ clsx( 'wp-block-playlist__tracklist', {
					'wp-block-playlist__tracklist-is-hidden': ! showTracklist,
					'wp-block-playlist__tracklist-artist-is-hidden':
						! showArtists,
				} ) }
```

scruffian replied:
done in 0938ec604e8

t-hamano commented on packages/block-library/src/playlist-track/edit.js:
```diff
@@ -0,0 +1,295 @@
+/**
+ * External dependencies
+ */
+import { v4 as uuid } from 'uuid';
```
Is there a reason why you didn't use `useInstanceId` hook?

scruffian replied:
I think this comment explains why:

> 	// Note that 'id' is the media attachment ID, while 'uniqueId' is a unique identifier.
>	// This is to make sure that the same media can be used in more than one track.

t-hamano commented on packages/block-library/src/playlist-track/edit.js:
```diff
@@ -0,0 +1,295 @@
+				/>
+			</BlockControls>
+			<InspectorControls>
+				<PanelBody title={ __( 'Settings' ) }>
```
Not a blocker, but it would be great to see this refactored to use ToolsPanel in the future. See https://github.com/WordPress/gutenberg/issues/67813

ntsekouras replied:
Yeah, all blocks should use `ToolsPanel` AFAIK.

t-hamano commented on packages/block-library/src/playlist-track/edit.js:
```diff
@@ -0,0 +1,295 @@
+						onChange={ ( value ) => {
+							setAttributes( { title: value } );
+						} }
+						keepPlaceholderOnFocus
```
There is no such prop in the RichText component.

scruffian replied:
removed in a43a4fd9ea9

t-hamano commented on packages/block-library/src/playlist-track/edit.js:
```diff
@@ -0,0 +1,295 @@
+							onChange={ ( value ) =>
+								setAttributes( { artist: value } )
+							}
+							keepPlaceholderOnFocus
```
There is no such prop in the RichText component.

t-hamano commented on packages/block-library/src/playlist-track/edit.js:
```diff
@@ -0,0 +1,295 @@
+						} }
+					/>
+					<MediaUploadCheck>
+						<div className="editor-video-poster-control">
```
Not a blocker: There is a reusable [PosterImage](https://github.com/WordPress/gutenberg/blob/80237f7256a86ad7656a660bdfdfaa15c59c12f4/packages/block-library/src/utils/poster-image.js#L31) component for poster images that we might be able to use.

<img width="341" height="176" alt="Image" src="https://github.com/user-attachments/assets/28b2886a-fbcb-4ff1-acaa-d3e7b6ba4f0b" />

t-hamano commented on packages/block-library/src/playlist-track/edit.js:
```diff
@@ -0,0 +1,295 @@
+					data-playlist-track-artist={ stripHTML( artist ) }
+					data-playlist-track-album={ stripHTML( album ) }
+					data-playlist-track-image-src={ image ?? null }
+					data-wp-context={ JSON.stringify( { uniqueId } ) }
```
Why are these attributes needed?

scruffian replied:
I don't think they are, I have removed them.

t-hamano commented on packages/block-library/src/playlist-track/index.php:
```diff
@@ -0,0 +1,74 @@
+
+	$wrapper_attributes = get_block_wrapper_attributes();
+
+	$unique_id = isset( $attributes['uniqueId'] ) ? $attributes['uniqueId'] : wp_generate_uuid4();
```
`wp_unique_id` might be more appropriate

scruffian replied:
changed

t-hamano commented on packages/block-library/src/playlist-track/index.php:
```diff
@@ -0,0 +1,74 @@
+		$html .= '<span class="wp-block-playlist-track__length">' .
+		sprintf(
+			/* translators: %s: track length in minutes:seconds */
+			'<span class="screen-reader-text">' . esc_html__( 'Length:' ) . ' </span>%s',
```
```suggestion
			'<span class="screen-reader-text">Length:</span>%s',
```

The text and placeholder order may vary depending on the locale.

t-hamano commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,469 @@
+						value={ order }
+						options={ [
+							{ label: __( 'Descending' ), value: 'DESC' },
+							{ label: __( 'Ascending' ), value: 'ASC' },
```
```suggestion
							{ label: __( 'Descending' ), value: 'desc' },
							{ label: __( 'Ascending' ), value: 'asc' },
```

In general, I think it's best to use lowercase values.

t-hamano reviewed (commented):
I've only seen the final code, but I'll comment on some points that I noticed. I apologize if I'm mistaken.

Also, I think it would be better to merge trunk into this branch again before shipping this PR. For example, [this if statement ](https://github.com/WordPress/gutenberg/pull/50664/files#diff-b26c6c8cc1dd9c9775a53b8f9fcc75cddd2a9a4d1a93846330662edce42e6141R246-R249)already exists in trunk, so this diff is weird.

scruffian committed:
Fix undefined SCSS variables in playlist blocks

Add missing @use import for base-styles/variables to resolve
build errors with $grid-unit-10 and $grid-unit-20 variables.

scruffian committed:
Fix array to string conversion error in blocks.php

The foreach loop was iterating over the array values instead of
keys, causing $blocks_dir to be an array instead of a string path.

scruffian committed:
Use trunk's simpler blocks directory structure

Replace the complex associative array structure with trunk's
simpler array of directory paths pointing to build/scripts/.

scruffian commented:
Feedback needed:

I have implemented WaveSurfer as a way of producing a visual representation of the audio file like this:

<img width="720" height="390" alt="Screenshot 2026-01-28 at 11 28 06" src="https://github.com/user-attachments/assets/19b0c814-9cff-420d-a77a-abfd9b5e0248" />

Getting WaveSurfer to work in the editor was really difficult because of the iframe. WaveSurfer itself relies on being at the document level and because the block is inside an iframe it can't access the DOM elements it needs in order to work. What I've done is it added an inline script which references WaveSurfer inside the iframe. This works but the problem is it uses an external CDN. We could update our webpack config so that the file comes from the local installation. However, before doing that, I wanted to get feedback on a few things:

1. Are we happy using WaveSurfer here? It's a very common library for this use. All the other libraries I've looked at seem to be built on top of it and therefore are likely to have similar problems with iframes.
2. Alternatively we could create our own solution which I started doing using Claude. I don't think this would take too long, but the complexity would mean we would end up duplicating a lot of logic and code from WaveSurfer into our own solution, which sounds like a maintenance nightmare to me.
3. Another option is simply to only use WaveSurfer on the front of the site and fall back to the standard HTML audio player in the editor. I'm reluctant to do that because we want the editor to match the front end.

MaggieCabrera commented:
Is it very complicated to make WaveSurfer work in the editor? I think it's fine to use an external utility. We have done this before with things like colors. Definitely no need to reinvent the wheel, even if we have AI to help

scruffian commented:
> Is it very complicated to make WaveSurfer work in the editor? I think it's fine to use an external utility. We have done this before with things like colors. Definitely no need to reinvent the wheel, even if we have AI to help

We would either need to use an external CDN or change our webpack config to serve WaveSurfer as a separate JS file.

MaggieCabrera commented:
> > Is it very complicated to make WaveSurfer work in the editor? I think it's fine to use an external utility. We have done this before with things like colors. Definitely no need to reinvent the wheel, even if we have AI to help
> 
> We would either need to use an external CDN or change our webpack config to serve WaveSurfer as a separate JS file.

I don't think the CDN is viable for GDPR reasons?

getdave commented:
Could the visual representation _in the Editor_ just be a placeholder? So it occupies the same space but doesn't have to be a literal waveform?

SirLouen reviewed (dismissed):
> I don't think the CDN is viable for GDPR reasons?

Using [jsdelivr is GDPR compliant](https://www.jsdelivr.com/terms/privacy-policy) 

Still this is the first CDN import in Gutenberg from `jsdelivr`

But still, I don't understand why adding it this way if its already packed. Am I missing something?

> I have implemented WaveSurfer as a way of producing a visual representation of the audio file like this:

@scruffian by the way, I don't feel that its a good idea to keep adding more features on top of this PR. I feel that its easier to review when the PR is shorter and goes straight to the point, and then future PR are added with additional features like this. This PR was already approved and ready to be merged before the new additions, it could have been merged, and then pushed a new PR with the new elements and keep iterating over that. 

I've blocked because new changes require new reviews, but feel free to unblock if its convenient.

scruffian committed:
replace wp_generate_uuid4 with wp_unique_id

scruffian committed:
Apply suggestion from @t-hamano

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

scruffian committed:
Apply suggestion from @getdave

Co-authored-by: Dave Smith <getdavemail@gmail.com>

scruffian committed:
make the placeholder part of the translatable section

scruffian committed:
remove invalid props

scruffian committed:
don't do string concat for classes

scruffian committed:
replace panelbody with toolspanel

scruffian committed:
remove unused attributes

scruffian committed:
make sorting use the track title

scruffian committed:
remove unneeded code

scruffian committed:
remove unneeded code

scruffian committed:
escape output

scruffian commented:
> I note that in the editor, clicking on the different tracks doesn't switch the heading values. In the front end it does. I'm not sure if this was expected or if it can be improved.

Yes I've noticed a few bugs in the editor - it doesn't play there either. My plan is to redo the player anyway, so I don't see much point in fixing it in this PR.

I have addressed all the comments on this PR. I am inclined to merge it and continue iterating in future PRs, since this one is already too big. I'll do that tomorrow unless someone finds more issues with this one.

t-hamano commented on packages/block-library/src/navigation/view.js.bak2:
Let's delete these temporary files 😄

scruffian replied:
Ooops! Done in https://github.com/WordPress/gutenberg/pull/75061
